packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -180,33 +195,46 @@ class RenderParagraph extends RenderBox {
     // us to validate our intrinsic sizes, which will change _textPainter's
     // layout because the intrinsic size calculations are destructive.
     final Size textSize = _textPainter.size;
-    size = constraints.constrain(textSize);
+    final double constrainedHeight =
+      _maxLines == null ? textSize.height :
+        textSize.height.clamp(0.0, _textPainter.preferredLineHeight * _maxLines);
```

Comment:
We probably want to teach it to stop after that number of lines.  Consider the case where the next line has a really tall glyph that overlaps the previous line.  If you clip here, the part of the glyph that overlaps the previous line will be drawn, which isn't what we want.